downstairs/src/extent.rs

```diff
@@ -0,0 +1,829 @@
+     */
+    pub async fn close(self) -> Result<(u64, u64, bool), CrucibleError> {
+        let inner = self.inner.lock().await;
+
+        let gen = inner.gen_number().unwrap();
+        let flush = inner.flush_number().unwrap();
+        let dirty = inner.dirty().unwrap();
+
+        Ok((gen, flush, dirty))
+    }
+
+    /**
+     * Create an extent at the location requested.
+     * Start off with the default meta data.
+     * Note that this function is not safe to run concurrently.
+     */
+    pub fn create(
+        dir: &Path,
+        def: &RegionDefinition,
+        number: u32,
+    ) -> Result<Extent> {
+        /*
+         * Store extent data in files within a directory hierarchy so that
+         * there are not too many files in any level of that hierarchy.
+         */
+        {
+            let path = extent_path(dir, number);
+
+            /*
+             * Verify there are not existing extent files.
+             */
+            if Path::new(&path).exists() {
+                bail!("Extent file already exists {:?}", path);
+            }
+        }
+        remove_copy_cleanup_dir(dir, number)?;
+
+        let inner = extent_inner_sqlite::SqliteInner::create(dir, def, number)?;
+
+        /*
+         * Complete the construction of our new extent
+         */
+        Ok(Extent {
+            number,
+            read_only: false,
+            iov_max: Extent::get_iov_max()?,
+            inner: Mutex::new(Box::new(inner)),
+        })
+    }
+
+    pub fn number(&self) -> u32 {
+        self.number
+    }
+
+    /// Read the real data off underlying storage, and get block metadata. If
+    /// an error occurs while processing any of the requests, the state of
+    /// `responses` is undefined.
+    #[instrument]
+    pub async fn read(
+        &self,
+        job_id: JobId,
+        requests: &[&crucible_protocol::ReadRequest],
+        responses: &mut Vec<crucible_protocol::ReadResponse>,
+    ) -> Result<(), CrucibleError> {
+        cdt::extent__read__start!(|| {
+            (job_id.0, self.number, requests.len() as u64)
+        });
+
+        let inner = self.inner.lock().await;
+
+        inner.read(job_id, requests, responses, self.iov_max)?;
+
+        cdt::extent__read__done!(|| {
+            (job_id.0, self.number, requests.len() as u64)
+        });
+
+        Ok(())
+    }
+
+    #[instrument]
+    pub async fn write(
+        &self,
+        job_id: JobId,
+        writes: &[&crucible_protocol::Write],
+        only_write_unwritten: bool,
+    ) -> Result<(), CrucibleError> {
+        if self.read_only {
+            crucible_bail!(ModifyingReadOnlyRegion);
+        }
+
+        cdt::extent__write__start!(|| {
+            (job_id.0, self.number, writes.len() as u64)
+        });
+
+        let mut inner = self.inner.lock().await;
+        inner.write(job_id, writes, only_write_unwritten, self.iov_max)?;
+
+        cdt::extent__write__file__done!(|| {
```

Comment:
This should move inside the sqlite specific impl

Reply:
Good catch, done in a9eee1a61fc4bb92588a7e1f2a69e10504f4db17